code/modules/mob/living/carbon/xenomorph/castes/widow/abilities_widow.dm

```diff
@@ -228,22 +162,175 @@
+/datum/action/xeno_action/attach_spiderlings
+	name = "Attach Spiderlings"
+	ability_name = "Attach Spiderlings"
+	mechanics_text = "Attach your current spiderlings to you "
+	action_icon_state = "attach_spiderling"
+	plasma_cost = 0
+	cooldown_timer = 0 SECONDS
+	keybind_signal = COMSIG_XENOABILITY_ATTACH_SPIDERLINGS
+	// the attached spiderlings
+	var/list/mob/living/carbon/xenomorph/spiderling/attached_spiderlings = list()
+	// how many times we attempt to attach adjacent spiderligns
+	var/attach_attempts = 5
+
+/datum/action/xeno_action/attach_spiderlings/action_activate()
+	. = ..()
+	var/mob/living/carbon/xenomorph/widow/X = owner
+	if(X.buckled_mobs)
+		/// yeet off all spiderlings if we are carrying any
+		X.unbuckle_all_mobs(TRUE)
+		return
+	var/datum/action/xeno_action/create_spiderling/create_spiderling_action = owner.actions_by_path[/datum/action/xeno_action/create_spiderling]
+	if(!(length(create_spiderling_action.spiderlings)))
+		X.balloon_alert(X, "No spiderlings")
+		return fail_activate()
+	var/list/mob/living/carbon/xenomorph/spiderling/remaining_spiderlings = create_spiderling_action.spiderlings.Copy()
+	grab_spiderlings(remaining_spiderlings, attach_attempts)
+	succeed_activate()
+
+// this proc scoops up adjacent spiderlings and then calls ride_widow on them
+/datum/action/xeno_action/attach_spiderlings/proc/grab_spiderlings(list/mob/living/carbon/xenomorph/spiderling/remaining_list, number_of_attempts_left)
+	var/mob/living/carbon/xenomorph/widow/X = owner
+	if(number_of_attempts_left <= 0)
+		return
+	for(var/mob/living/carbon/xenomorph/spiderling/remaining_spiderling AS in remaining_list)
+		if(!X.Adjacent(remaining_spiderling))
+			continue
+		remaining_list -= remaining_spiderling
+		X.buckle_mob(remaining_spiderling, TRUE, TRUE, 90, 1,0)
+	addtimer(CALLBACK(src, .proc/grab_spiderlings, remaining_list, number_of_attempts_left - 1), 1)
+
+// ***************************************
+// *********** Web Hook
+// ***************************************
+/datum/action/xeno_action/activable/web_hook
+	name = "Web Hook"
+	ability_name = "Web Hook"
+	mechanics_text = "Shoot out a web and pull it to traverse forward"
+	action_icon_state = "web_hook"
+	plasma_cost = 50
+	cooldown_timer = 20 SECONDS
+	keybind_signal = COMSIG_XENOABILITY_WEB_HOOK
+	//ref to beam for web hook
+	var/datum/beam/web_beam
+	//Do travel the full distnace or half?
+	var/full_distance = TRUE
+
+/datum/action/xeno_action/activable/web_hook/can_use_ability(atom/A)
+	. = ..()
+	if(isliving(A))
+		return FALSE
+	if(!isturf(A))
+		return FALSE
+	var/turf/current = get_turf(owner)
+	var/turf/target_turf = get_turf(A)
+	if(get_dist(current, target_turf) > WIDOW_WEB_HOOK_RANGE)
+		return FALSE
+	current = get_step_towards(current, target_turf)
+	while((current != target_turf))
+		if(current.density)
+			full_distance = TRUE
+			return TRUE
+		current = get_step_towards(current, target_turf)
+	if(!current.density)
+		full_distance = FALSE
```

Comment:
1) What if you target the turf you are on
2) you convey meaning better if you check target_turf density at the end, no current density
3) I don't know what are you even doing, why is it returning . and not true directly. That makes the whole thing confusing
4) You need to check if . is not already false before doing your checks